runtime/src/main/java/org/corfudb/runtime/clients/NettyClientProtobufRouter.java

```diff
@@ -0,0 +1,360 @@
+     * @param client The client to add to the router
+     * @return This IClientProtobufRouter
+     */
+    public IClientProtobufRouter addClient(IClient client) {
+        // TODO: Rewrite IClient and corresponding methods,
+
+        return this;
+    }
+
+    /**
+     * Send a request message and get a completable future to be fulfilled by the reply.
+     *
+     * @param request The request message to send.
+     * @param ctx Context of the channel handler.
+     * @param <T> The type of completable to return.
+     * @return A completable future which will be fulfilled by the reply,
+     * or a timeout in the case there is no response.
+     */
+    @Override
+    public  <T> CompletableFuture<T> sendRequestAndGetCompletable(Request request, ChannelHandlerContext ctx) {
+
+        // Check the connection future. If connected, continue with sending the message.
+        // If timed out, return a exceptionally completed with the timeout.
+        try {
+            connectionFuture
+                    .get(parameters.getConnectionTimeout().toMillis(), TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            throw new UnrecoverableCorfuInterruptedError(e);
+        } catch (TimeoutException te) {
+            CompletableFuture<T> f = new CompletableFuture<>();
+            f.completeExceptionally(te);
+            return f;
+        } catch (ExecutionException ee) {
+            CompletableFuture<T> f = new CompletableFuture<>();
+            f.completeExceptionally(ee.getCause());
+            return f;
+        }
+
+        // Set up the timer and context to measure request
+        final Timer roundTripMsgTimer = CorfuRuntime.getDefaultMetrics()
+                .timer(timerNameCache.get(message.getMsgType()));
+
+        final Timer.Context roundTripMsgContext = MetricsUtils
+                .getConditionalContext(roundTripMsgTimer);
+
+        // Get the next request ID
+        final long thisRequest = requestID.getAndIncrement();
+
+        // Set the base fields for this message.
+        Request.Builder reqBuilder = request.toBuilder();
+        reqBuilder.getHeaderBuilder().setClientId(parameters.getClientId()).setRequestId(thisRequest);
+        request = reqBuilder.build();
+
+        // Generate a future and put it in the completion table.
+        final CompletableFuture<T> cf = new CompletableFuture<>();
+        outstandingRequests.put(thisRequest, cf);
+
+        // Write this message out on the channel
+        ByteBuf outBuf = PooledByteBufAllocator.DEFAULT.buffer();
+        ByteBufOutputStream requestOutputStream = new ByteBufOutputStream(outBuf);
+
+        try {
+            // Mark this message as Protobuf message (temporarily)
+            requestOutputStream.writeByte(API.PROTO_CORFU_MSG_MARK);
+            request.writeTo(requestOutputStream);
+            ctx.writeAndFlush(outBuf, ctx.voidPromise());
+            log.trace("Sent one-way request message: {}", request.getHeader());
```

Comment:
This is not a one-way message.

Reply:
I will correct it in next commit. Thanks!